python/lsst/daf/butler/datastores/fileLikeDatastore.py

```diff
@@ -825,7 +830,69 @@ def _write_in_memory_to_artifact(self, inMemoryDataset: Any, ref: DatasetRef) ->
         info : `StoredFileInfo`
             Information describin the artifact written to the datastore.
         """
-        raise NotImplementedError()
+        location, formatter = self._prepare_for_put(inMemoryDataset, ref)
+        uri = location.uri
+
+        if uri.exists():
+            # Assume that by this point if registry thinks the file should
+            # not exist then the file should not exist and therefore we can
+            # overwrite it. This can happen if a put was interrupted by
+            # an external interrupt. The only time this could be problematic is
+            # if the file template is incomplete and multiple dataset refs
+            # result in identical filenames.
+            # Eventually we should remove the check completely (it takes
+            # non-zero time for network).
+            log.warning("Object %s exists in datastore for ref %s", uri, ref)
+
+        if not uri.dirname().exists():
+            log.debug("Folder %s does not exist yet so creating it.", uri.dirname())
+            uri.dirname().mkdir()
+
+        if self._transaction is None:
+            raise RuntimeError("Attempting to write artifact without transaction enabled")
+
+        def _removeFileExists(uri: ButlerURI) -> None:
+            """Remove a file and do not complain if it is not there.
+
+            This is important since a formatter might fail before the file
+            is written and we should not confuse people by writing spurious
+            error messages to the log.
+            """
+            try:
+                uri.remove()
+            except FileNotFoundError:
+                pass
+
+        # Register a callback to try to delete the uploaded data if
+        # something fails below
+        self._transaction.registerUndo("artifactWrite", _removeFileExists, uri)
+
+        # For a local file, simply use the formatter directly
+        if uri.isLocal:
+            path = formatter.write(inMemoryDataset)
+            assert self.root.join(path) == uri
+            log.debug("Successfully wrote python object to local file at %s", uri)
+        else:
+            # This is a remote URI, so first try bytes and write directly else
+            # fallback to a temporary file
+            try:
+                serializedDataset = formatter.toBytes(inMemoryDataset)
+                log.debug("Writing bytes directly to %s", uri)
+                uri.write(serializedDataset, overwrite=True)
+                log.debug("Successfully wrote bytes directly to %s", uri)
+            except NotImplementedError:
+                with tempfile.NamedTemporaryFile(suffix=uri.getExtension()) as tmpFile:
+                    # Need to configure the formatter to write to a different
+                    # location and that needs us to overwrite internals
+                    tmpLocation = Location(*os.path.split(tmpFile.name))
+                    formatter._fileDescriptor.location = tmpLocation
```

Comment:
Could we make this mutation a public method on the `Formatter` class without modifying all of its derived classes?

Reply:
I can add a quick `Formatter.updateLocation` method although maybe with an underscore for now.